src/merkle/tiered_smt/storage.rs

```diff
@@ -0,0 +1,150 @@
+use super::{BTreeMap, CanonicalWord, NodeIndex, NodeType, Vec, Word};
+use core::convert::Infallible;
+
+/// In-memory storage is infallible.
+pub type StorageError = Infallible;
+
+// TIERED SPARSE MERKLE TREE STORAGE
+// ================================================================================================
+
+/// A storage implementation for the tiered sparse merkle tree.
+#[derive(Debug, Default)]
+pub struct Storage {
+    types: BTreeMap<NodeIndex, NodeType>,
+    nodes: BTreeMap<NodeIndex, Word>,
+    keys: BTreeMap<CanonicalWord, NodeIndex>,
+    upper_leaf_keys: BTreeMap<NodeIndex, CanonicalWord>,
+    leaf_values: BTreeMap<CanonicalWord, Word>,
+    ordered_leaves: BTreeMap<u64, Vec<CanonicalWord>>,
+}
+
+impl Storage {
+    // PROVIDERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the type of a node.
+    pub fn get_type(&self, index: &NodeIndex) -> Result<Option<NodeType>, StorageError> {
+        Ok(self.types.get(index).copied())
+    }
+
+    /// Returns the value of a node.
+    pub fn get_node(&self, index: &NodeIndex) -> Result<Option<Word>, StorageError> {
+        Ok(self.nodes.get(index).copied())
+    }
+
+    /// Returns the index of a leaf key.
+    pub fn get_leaf_index(&self, key: &CanonicalWord) -> Result<Option<NodeIndex>, StorageError> {
+        Ok(self.keys.get(key).copied())
+    }
```

Comment:
I would change the signature here to:
```Rust
pub fn get_leaf_index(&self, key: u64) -> Result<NodeIndex, MerkleError>
```
I think we should be able to implement this method with several lookups.